src/renderers/webgl/WebGLBindingStates.js

```diff
@@ -176,6 +176,8 @@
 			const cachedAttribute = cachedAttributes[ key ];
 			const geometryAttribute = geometryAttributes[ key ];
 
+			if ( geometryAttribute.isGLBufferAttribute ) return true;
```

Comment:
Can you please explain why you have added this line? When removed, the respective demo does not seem to work correctly anymore. 

I don't think it's correct to always force a call of `setupVertexAttributes()` on each frame.

Reply:
I don't exactly understand what this part of the system is supposed to do. But my guess is:

- if there is no VAO, bind appropriate buffers
- if there is a VAO and the attributes were not changed, just switch to that VAO
- if attributes changed, update the VAO, i.e. switch to it and bind the updated buffers

That all sounds good in theory. But when I noticed my demo was broken, I started debugging and ended up here. So basically I set `needsUpdate` to my attribute and expect the engine to do the job of updating whatever was changed (the buffer). That was not the case here.

For some reason, `cachedAttribute` always contained an already-updated version of my attribute. So the VAO never got updated (hence the demo was broken). I double-checked this fact by adding some custom properties directly into the buffer, like `buffer.name=111`.

So I decided to opt-out from this optimization with my attribute type. I expect that line 179 effectively makes VAO unused for `GLBufferAttribute`. That is fine with me. Although if you understand what is going wrong and how to make it work with this optimization, that would also be great.

Reply:
The problem is that the current implementation is bad for performance since a lot of unnecessary state changes are done. 

We definitely have to revisit this issue otherwise I would vote to revert `GLBufferAttribute` before the next release.

Reply:
The current implementation is vastly superior than having no implementation. And if I'm right, for `webgl1` people the further discussion makes no sense, since they always get exactly this level of performance. So there is a considerable group of people who either get this feature or not, regardless of VAO optimizations.

Now to optimizations. I would gladly fixed the VAO issue, given it was at all possible for you (or anybody) to explan why the **cached** attribute is always the updated attribute itself. And if it is the correct behavior, how it can be possibly fixed.

If you want, I might go on and report this behavior as a bug in a separate issue. Because now it is my understanding, that when I set `needsUpdate` on my attribute, the cache should contain the old version. If I have just updated the attribute, and the "cached" version does already contain what I have updated, **it is a bug in WebGLBindingStates** unless proven otherwise.

Reply:
This has nothing to do with WebGL 1 or 2. VAO is available for most WebGL 1 users. Besides, even without VAO the buffer attribute bindings should only be done if necessary. Not if just the contents of a buffer changes.

I think the problem in your demo is this line:
```js
posAttr.buffer = ( posAttr.buffer === pos ) ? pos2 : pos;
```
You should exchange the entire attribute and not just the buffer. Or you keep the buffer the same and only change the contents.

Reply:
This pattern works without the hack in `WebGLBindingStates`:
```js
var posAttr1 = new THREE.GLBufferAttribute( pos, gl.FLOAT, 3, 4, particles );
var posAttr2 = new THREE.GLBufferAttribute( pos2, gl.FLOAT, 3, 4, particles );
geometry.setAttribute( 'position', posAttr1 );

setInterval( function () {

    var attr = geometry.getAttribute( 'position' );

    geometry.setAttribute( 'position', ( attr === posAttr1 ) ? posAttr2 : posAttr1 );
 
}, 2000 );
```